File: CSES/11. Additional Problems/CSES-1706 School Excursion.py
class UnionFind:
    def __init__(self, n):
        self.pa = [i for i in range(n)]
        self.sz = [1] * n

    def find(self, x):
        if self.pa[x] != x:
            self.pa[x] = self.find(self.pa[x])
        return self.pa[x]

    def union(self, x, y):
        fx, fy = self.find(x), self.find(y)
        if fx == fy:
            return
        if self.sz[fx] > self.sz[fy]:
            fx, fy = fy, fx
        self.pa[fx] = fy
        self.sz[fy] += self.sz[fx]

n, m = map(int, input().split())

uf = UnionFind(n)
for _ in range(m):
    u, v = map(lambda x: int(x) - 1, input().split())
    uf.union(u, v)

arr = []
vis = [False] * n
for x in range(n):
    fx = uf.find(x)
    if vis[fx]:
        continue
    vis[fx] = True
    arr.append(uf.sz[fx])

# f = [False] * (n + 1)
# f[0] = True
# for x in arr:
#     for i in range(n, x - 1, -1):
#         f[i] |= f[i - x]

f = 1
for x in arr:
    f |= f << x

for i in range(1, n + 1):
    print("1" if f >> i & 1 else "0", end="")
print()

File: CSES/18. Additional Problems II/CSES-1706 School Excursion.cpp
#include <bits/stdc++.h>
using namespace std;
const int N = 1e5 + 5;
#define endl '\n'

class UnionFind {
public:
    vector<int> pa, sz;
    UnionFind(int n) : sz(n, 1), pa(n) {
        iota(pa.begin(), pa.end(), 0);
    }

    int find(int x) {
        return pa[x] == x ? x : pa[x] = find(pa[x]);
    }

    void merge(int x, int y) {
        int fx = find(x), fy = find(y);
        if (fx == fy) return;
        if (sz[fx] > sz[fy]) swap(fx, fy);
        pa[fx] = fy;
        sz[fy] += sz[fx];
    }
};

int main() {
    ios::sync_with_stdio(false); cin.tie(nullptr); cout.tie(nullptr);
    int n, m, u, v;
    cin >> n >> m;
    UnionFind uf(n);
    for (int i = 0; i < m; i++) {
        cin >> u >> v;
        uf.merge(u - 1, v - 1);
    }
    vector<int> arr;
    vector<bool> vis(n);
    for (int i = 0; i < n; i++) {
        int fx = uf.find(i);
        if (vis[fx]) continue;
        vis[fx] = true;
        arr.push_back(uf.sz[fx]);
    }
    bitset<N> f;
    f.set(0);
    for (int x : arr) {
        f |= f << x;
    }
    for (int i = 1; i <= n; i++) cout << f[i];
    cout << endl;
    return 0;
}